### DISCUSSION TITLE
How does `-o modules` work?

### DISCUSSION THREAD

**Discussion by sharadhr** [libfuse/libfuse#891] [Q&A]
I've been using `libfuse` (specifically, libfuse 2.9.9) in a personal application by making changes to `example/fusexmp_fh.c`, and I'm _statically_ linking to `libfuse.a`.

I'd like to support the `-o modules` option, but when I try to pass this to my program, it crashes with the following error message:

```
fuse: dlopen(libfusemod_subdir.so, 0x0002): tried: 'libfusemod_subdir.so' (no such file)
```

I've looked at my distribution's file list for fuse2, and I don't see `libfusemod_subdir.so` anywhere. In fact I've checked several other distributions, and none of them appear to distribute any `libfusemod_%s.so` files at all. 

I understand the `subdir` implementation lives in `modules/subdir.c`, and the above error comes from failing to load the shared object in the following line: https://github.com/libfuse/libfuse/blob/d04687923194d906fe5ad82dcd546c9807bf15b6/lib/fuse.c#L239

How can I fix this, and where is the submodule expected to be found? Am I misunderstanding how FUSE modules work in the first place? I'm struggling to find any documentation on how this works, and `libfusemod` turns up very few results on a Web search beyond the code itself.

**Comment by bsbernd** [comment]
I didn't check libfuse2 and I'm also not terribly familiar with modules, but subdir and iconv (if available) are *built-in* modules. No need to load it with -omodule=subdir. In libfuse3, see `example/passthrough -h`. You should see a similar help text with your libfuse2 binary

```
Options for subdir module:
    -o subdir=DIR           prepend this directory to all paths (mandatory)
    -o [no]rellinks         transform absolute symlinks to relative

Options for iconv module:
    -o from_code=CHARSET   original encoding of file names (default: UTF-8)
    -o to_code=CHARSET     new encoding of the file names (default: UTF-8)
```

I.e. these built-in modules are already loaded, no point to load them again and as you noticed, there is also .so lib for them.

**Comment by trapexit** [comment]
I'd argue that that feature of libfuse should never have existed (along with the arg parsing) and you would be better served by adding such behavior to your own project. You'll have more control.

**Reply by bsbernd** [reply]
Yeah, I wish so, but it dates back to 2007. At best it could have been removed with libfuse3.
